### PR TITLE
o/devicestate: use device-service.access to determine if we should fetch a serial or not

### PR DESCRIPTION
**Pull request by andrewphelpsj**
Use a combination of `device-service.access`, `device-service.url`, and `store.access` to determine if we should fetch a device serial or not. The logic goes as follows:
* If we have a gadget:
  * `device-service.access=offline`: do not fetch device serial
  * `device-service.url` is unset (we'd be using the default store): use `store.access` to determine if we should fetch a device serial or not.
  * Any other case: do fetch device serial
* If we don't have a gadget:
  * Use `store.access` to determine if we should fetch a device serial or not

This feels a little complicated, so if there is an easier way of expressing this logic, please let me know.

### CODE REVIEW

File: overlord/devicestate/handlers_serial.go [hunk 1]
```diff
@@ -679,6 +689,46 @@ func getSerialRequestConfig(t *state.Task, regCtx registrationContext, client *h
...
+
+	var deviceServiceURL string
+	if err := tr.GetMaybe(gadgetName, "device-service.url", &deviceServiceURL); err != nil {
+		return false, err
+	}
+
+	// if we already have a device serial and we'd be using the fallback
+	// device-service.url (which is the store), then use store.access to
+	// determine if we should request
+	if deviceSerial != "" && deviceServiceURL == "" {
```
Comment (pedronis) [inline overlord/devicestate/handlers_serial.go:725]: I don't understand why it matters whether we have a serial or not, we shouldn't do requestes if the url is ""  and storreAccess is offline

**Review by pedronis — CHANGES_REQUESTED** [review]
thanks, did a pass, the checks on serial are confusing, don't think we want them

File: overlord/devicestate/handlers_serial.go (hunk 1 above)
Comment (andrewphelpsj) [reply overlord/devicestate/handlers_serial.go]: Was a misunderstanding on my part, fixed now. Thanks!

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -611,6 +611,19 @@ func (m *DeviceManager) ensureOperational() error {
 		return fmt.Errorf("internal error: core device brand and model are set but there is no model assertion")
 	}
 
+	willRequestSerial, err := shouldRequestSerial(m.state, gadget, device.Serial)
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/devicestate/devicemgr.go]: Fixed, thanks!

File: overlord/devicestate/devicestate_serial_test.go [hunk 3]
```diff
@@ -2408,3 +2409,192 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
...
+	s.state.Unlock()
+
+	err := s.mgr.Ensure()
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	tasks := s.state.Tasks()
+	s.state.Unlock()
+
+	sort.Slice(tasks, func(l, r int) bool {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/devicestate/devicestate_serial_test.go:2520]: the source of Tasks is a map so there's not predefinite order, anyway checking for the full Tasks is unusual, what tends here tends to do is find the expected Change and look at that, see the findBecomeOperationalChange helper here I think

File: overlord/devicestate/handlers_serial.go [hunk 4]
```diff
@@ -513,6 +513,16 @@ func getSerial(t *state.Task, regCtx registrationContext, privKey asserts.Privat
...
 
 	st := t.State()
+
+	shouldRequest, err := shouldRequestSerial(st, regCtx.GadgetForSerialRequestConfig())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if !shouldRequest {
+		return nil, nil, nil
```
Comment (pedronis) [inline overlord/devicestate/handlers_serial.go:523]: this should do t.Logf though as I asked

**Review by pedronis — COMMENTED** [review]
thanks, did another pass

File: overlord/devicestate/devicestate_serial_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/devicestate/devicestate_serial_test.go:2520]: In this situation, I'm conditionally adding tasks. Should I use a function like that to look for the change, and then look at the tasks on that change?

File: overlord/devicestate/handlers_serial.go (hunk 4 above)
Comment (andrewphelpsj) [reply overlord/devicestate/handlers_serial.go:524]: Ugh, sorry. Not sure what happened there.

File: overlord/devicestate/devicestate_serial_test.go [hunk 5]
```diff
@@ -2408,3 +2409,193 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
...
+	c.Assert(err, IsNil)
+	tr.Commit()
+
+	s.state.Unlock()
+
+	err = s.mgr.Ensure()
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	tasks := s.state.Tasks()
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/devicestate/devicestate_serial_test.go]: Done!

File: overlord/devicestate/devicestate_serial_test.go [hunk 6]
```diff
@@ -2408,3 +2409,193 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
...
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/devicestate/devicestate_serial_test.go]: Could you explain the reason to use one over the other, when you get a chance? Is it just so that the tests are more isolated, and something else adding a task would not break them?

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/devicestate/devicestate_serial_test.go [hunk 7]
```diff
@@ -2408,3 +2409,194 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
...
+			err = tr.Set(t.gadgetName, "device-service.access", t.deviceServiceAccess)
+			c.Assert(err, IsNil)
+			err = tr.Set(t.gadgetName, "device-service.url", t.deviceServiceURL)
+			c.Assert(err, IsNil)
+		}
+		tr.Commit()
+
+		shouldReqest, err := devicestate.ShouldRequestSerial(s.state, t.gadgetName)
+		c.Check(err, IsNil)
+		c.Check(shouldReqest, Equals, t.expected, Commentf("%+v", t))
```
Comment (MiguelPires) [inline overlord/devicestate/devicestate_serial_test.go:2492]:
```suggestion
		shouldRequest, err := devicestate.ShouldRequestSerial(s.state, t.gadgetName)
		c.Check(err, IsNil)
		c.Check(shouldRequest, Equals, t.expected, Commentf("%+v", t))
```

Comment (MiguelPires) [inline overlord/devicestate/devicestate_serial_test.go:2492]: If this fails, the output will be hard to parse. It would be easier to understand if the test cases had a name that we could use to print and match to a test. If the test names would be too awkward or hard to come up with, then just printing the index of the failed tests would be fine

File: overlord/devicestate/devicestate_serial_test.go [hunk 8]
```diff
@@ -2408,3 +2409,194 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
...
+
+	err := s.mgr.Ensure()
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	change := s.findBecomeOperationalChange()
+	tasks := change.Tasks()
+	s.state.Unlock()
+
+	sort.Slice(tasks, func(l, r int) bool {
```
Comment (MiguelPires) [inline overlord/devicestate/devicestate_serial_test.go:2521]: There's a `testutil.DeepUnsortedMatches` that can compare these things without having to sort them

**Review by MiguelPires — APPROVED** [review]
Sorry for the delay. LGTM, thanks

File: overlord/devicestate/devicestate_serial_test.go (hunk 8 above)
Comment (andrewphelpsj) [reply overlord/devicestate/devicestate_serial_test.go:2521]: Thanks, that is useful to know! In this case though, I'd just need to verify the kind of task.
